man/join1.1
.TH  Tukubai 1 "19 Jun 2022" "usp Tukubai" "Tukubai コマンド マニュアル"

.br
join1(1)
.br

.br
【名前】
.br

.br
join1\ :\ トランザクションファイルにマスタファイルを連結(一致した行のみ連結)
.br

.br
【書式】
.br

.br
Usage\ \ \ :\ join1\ [+ng[<fd>]]\ key=<n>\ <master>\ [<tran>]
.br
Version\ :\ Sun\ Jun\ 19\ 23:55:51\ JST\ 2022
.br
Edition\ :\ 1
.br

.br
【説明】
.br

.br
tranのkey=<n>で指定したキーフィールドがmasterの第1フィールド(キーフィールド)と
.br
一致した行のみtranから抽出して、masterの情報を連結して出力する。連結はtranのキ
.br
ーフィールドの直後にmasterの内容を挿入連結する形で実施される。
.br

.br
masterに-を指定すると標準入力をマスタファイルとする。tranが無指定かまたは-が指
.br
定されている場合には、標準入力がトランザクションファイルとなる。
.br

.br
masterの第1フィールドおよびtranの第<n>フィールドは必ず昇順で整列されていること
.br
が条件となる。また、masterのキーフィールド(第1フィールド)はユニークでなければな
.br
らない。(第1フィールドが同じ値をもつ行が複数あってはならない)。tranについてはこ
.br
の制約はなく、キーフィールド(第<n>フィールド)が同じ値の行はいくつあっても構わな
.br
い。
.br

.br
キー指定にひとつでも0があると全フィールドを指定したことになる
.br

.br
+ng<fd>オプションを使うことで、一致した行を標準出力ファイルへ、一致しなかった行
.br
をファイルデスクリプタ<fd>のファイルへ出力することが可能。<fd>を省略した場合は
.br
標準エラー出力へ出力する。
.br

.br
tranでキーが整列されていないものはjoin1(1)では処理できない。その場合は一旦tran
.br
を整列してからjoin1(1)で処理するか、join1(1)のかわりにcjoin1(1)を使って処理すれ
.br
ばよい。
.br

.br
【例1】基本パターン
.br

.br
経費ファイルkeihiからmasterに登録されている4人のデータを抽出し、マスタファイル
.br
の内容を連結する。
.br

.br

  $ cat master
  0000003 杉山______ 26 F
  0000005 崎村______ 50 F
  0000007 梶川______ 42 F
  0000010 柳本______ 50 F

  $

.br

  $ cat keihi
  20070401 0000001 300
  20070403 0000001 500
  20070404 0000001 700
  20070401 0000003 200
  20070402 0000003 400
  20070405 0000003 600
  20070401 0000005 250
  20070402 0000005 450
  20070402 0000007 210
  20070404 0000007 410
  20070406 0000007 610

  $

.br
keihiの第2フィールドをキーにしてmasterを結合する。masterに存在するキーをもつ行
.br
にしか連結を行わない。存在しない行は捨てられる。keihiは同じキーを持つ行が複数行
.br
あっても構わない。
.br

.br

  $ join1 key=2 master keihi
  20070401 0000003 杉山______ 26 F 200
  20070402 0000003 杉山______ 26 F 400
  20070405 0000003 杉山______ 26 F 600
  20070401 0000005 崎村______ 50 F 250
  20070402 0000005 崎村______ 50 F 450
  20070402 0000007 梶川______ 42 F 210
  20070404 0000007 梶川______ 42 F 410
  20070406 0000007 梶川______ 42 F 610

  $

.br
【例2】
.br

.br
左から順に連続した複数のフィールドをキーに指定する場合は次のように実行する。
.br
kekkaファイルの第2フィールド、第2フィールドをキーにして(第2、第3フィールド順に
.br
整列されている)masterファイル(第1、第2フィールドの順に整列されている)に存在する
.br
行のみ抽出して、masterの内容を連結して出力している。
.br

.br

  $ cat master
  A 0000003 杉山______ 26 F
  A 0000005 崎村______ 50 F
  B 0000007 梶川______ 42 F
  C 0000010 柳本______ 50 F

  $

.br

  $ cat kekka
  1 A 0000000 91 59 20 76 54
  2 A 0000001 46 39 8  5  21
  3 A 0000003 30 50 71 36 30
  4 A 0000004 58 71 20 10 6
  5 A 0000005 82 79 16 21 80
  6 B 0000007 50 2  33 15 62
  7 B 0000008 52 91 44 9  0
  8 C 0000009 60 89 33 18 6
  9 C 0000010 95 60 35 93 76
  10 C 0000011 92 56 83 96 75

  $

.br

  $ join1 key=2/3 master kekka
  3 A 0000003 杉山______ 26 F 30 50 71 36 30
  5 A 0000005 崎村______ 50 F 82 79 16 21 80
  6 B 0000007 梶川______ 42 F 50 2  33 15 62
  9 C 0000010 柳本______ 50 F 95 60 35 93 76

  $

.br
左から順に連続していない複数のフィールドをキーに指定することも可能。この場合は
.br
複数のキーフィールドを@でつなげて指定する。なお次の例の場合には、tranは第4、第2
.br
フィールドの順に整列され、masterは第3、第1フィールドの順に整列されていることが
.br
必要となる。
.br

.br

  $ join1 key=4@2 master tran

.br
【例3】"+ng"\ オプション
.br

.br
masterとキーに一致しない行を抽出することも可能。キーに一致する行は標準出力に、
.br
一致しない行は標準エラー出力に出力される。この場合、一致する行はマスタファイル
.br
と連結されるが、一致しない行はマスタファイルに存在しないため連結されずそのまま
.br
出力されることになる。
.br

.br

  $ cat master
  0000003 杉山______ 26 F
  0000005 崎村______ 50 F
  0000007 梶川______ 42 F
  0000010 柳本______ 50 F

  $

.br

  $ cat kekka
  0000000 91 59 20 76 54
  0000001 46 39 8  5  21
  0000003 30 50 71 36 30
  0000004 58 71 20 10 6
  0000005 82 79 16 21 80
  0000007 50 2  33 15 62
  0000008 52 91 44 9  0
  0000009 60 89 33 18 6
  0000010 95 60 35 93 76
  0000011 92 56 83 96 75

  $

.br
成績ファイルkekkaからmasterに登録されている4名のデータとその他のデータはそれぞ
.br
れ次のように抽出できる。
.br

.br

  $ join1 +ng key=1 master kekka > ok-data 2> ng-data

.br

  $ cat ok-data           ←　一致したデータ
  0000003 杉山______ 26 F 30 50 71 36 30
  0000005 崎村______ 50 F 82 79 16 21 80
  0000007 梶川______ 42 F 50 2  33 15 62
  0000010 柳本______ 50 F 95 60 35 93 76

  $

.br

  $ cat ng-data           ←　一致しなかったデータ
  0000000 91 59 20 76 54
  0000001 46 39 8  5  21
  0000004 58 71 20 10 6
  0000008 52 91 44 9  0
  0000009 60 89 33 18 6
  0000011 92 56 83 96 75

  $

.br
【関連項目】
.br

.br
cjoin0(1)、cjoin1(1)、cjoin2(1)、join0(1)、join2(1)、loopj(1)、loopx(1)、up3(1)
.br
、マスタファイル(5)、トランザクションファイル(5)
.br

.br
last\ modified:\ Fri\ Jul\ 15\ 15:42:38\ JST\ 2022
.br
Contact\ us:\ uecinfo@usp-lab.com
.br
Copyright\ (c)\ 2012-2022\ Universal\ Shell\ Programming\ Laboratory\ All\ Rights
.br


COMMANDS.HS/join1.hs
#!/usr/bin/env runghc --
import System.Environment
import System.IO
import Text.ParserCombinators.Parsec
import Control.Monad
import Data.ByteString.Lazy.Char8 as BS hiding (length,take,drop,filter,head)
import Control.Applicative hiding ((<|>), many)

{--
join1（Open usp Tukubai）

designed by Nobuaki Tounaka
written  by Hinata Yanagi

The MIT License

Copyright (C) 2022 Universal Shell Programming Laboratory

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
--}

showUsage :: IO ()
showUsage = do
    System.IO.hPutStr stderr "Usage    : join1 [+ng] <key=n> <master> [<tran>]\n"
    System.IO.hPutStr stderr "Version  : Sat Oct  1 16:50:42 JST 2022\n"
    System.IO.hPutStr stderr "Open usp Tukubai (LINUX+FREEBSD)\n"

main :: IO ()
main = do args <- getArgs
          case args of
                []            -> showUsage
                ["-h"]        -> showUsage
                ["--help"]    -> showUsage
                ["--version"] -> showUsage
                ["+ng",key,master,tran] -> mainProc True  key master tran
                ["+ng",key,master]      -> mainProc True  key master "-"
                [key,master,tran]       -> mainProc False key master tran
                [key,master]            -> mainProc False key master "-"
                _                       -> showUsage


mainProc :: Bool -> String -> String -> String -> IO ()
mainProc ng key master tran = do ms <- readF master
                                 ts <- readF tran
                                 mainProc' ng (parseKey key) ms ts

readF :: String -> IO BS.ByteString
readF "-" = BS.getContents
readF f   = BS.readFile f

parseKey :: String -> Keys
parseKey str = case parse keys "" str of
                    Right opt -> opt
                    Left  err -> Error (show err)

mainProc' :: Bool -> Keys -> BS.ByteString -> BS.ByteString -> IO ()
mainProc' ng (Keys ks) ms ts = out ng (join1 ng (head mlines) (drop 1 mlines) tlines) 
                               where mlines = parseMaster ks (BS.lines ms)
                                     tlines = parseTran ks (BS.lines ts)

out :: Bool -> [OutTran] -> IO ()
out _  []                  = do return ()
out False ((OkTran ln):as) = (BS.putStrLn ln) >> (out False as)
out True  ((OkTran ln):as) = (BS.putStrLn ln) >> (out True  as)
out False ((NgTran ln):as) = out False as
out True  ((NgTran ln):as) = (BS.hPutStrLn stderr ln) >> (out True as)

join1 :: Bool -> Master -> [Master] -> [Tran] -> [OutTran]
join1 _ _ _ []    = []
join1 ng (Master mk v) [] ((Tran p tk a):ts) 
  | mk == tk      = OkTran (toStr2 v (Tran p tk a)) : join1 ng (Master mk v) [] ts
  | ng == True    = NgTran (toStr (Tran p tk a)) : join1 ng (Master mk v) [] ts
  | otherwise     = join1 ng (Master mk v) [] ts
join1 ng (Master mk v) (m:ms) ((Tran p tk a):ts) 
  | mk == tk      = OkTran (toStr2 v (Tran p tk a)) : join1 ng (Master mk v) (m:ms) ts 
  | mk <  tk      = join1 ng m ms ((Tran p tk a):ts) 
  | ng == True    = NgTran (toStr (Tran p tk a)) : join1 ng (Master mk v) (m:ms) ts
  | otherwise     = join1 ng (Master mk v) (m:ms) ts

toStr (Tran p k a) = BS.unwords $ filter (/= (BS.pack "")) [p,k,a]
toStr2 v (Tran p k a) = BS.unwords $ filter (/= (BS.pack "")) [p,k,v,a]

makeLine :: Maybe Master -> Tran -> OutTran
makeLine (Just (Master k v)) (Tran p _ a) = OkTran (BS.unwords [p,k,a])
makeLine Nothing             (Tran p k a) = NgTran (BS.unwords [p,k,a])

myWords :: BS.ByteString -> [BS.ByteString]
myWords line = BS.split ' ' line

matchMaster k (Master a b) = k == a
              
parseMaster :: [Int] -> [BS.ByteString] -> [Master]
parseMaster ks lines = [ f (length ks) (myWords ln) | ln <- lines ]
                       where f n ws = Master (BS.unwords $ take n ws) (BS.unwords $ drop n ws)

parseTran :: [Int] -> [BS.ByteString] -> [Tran]
parseTran ks lines = [ parseTran' ks (myWords ln) | ln <- lines ]

parseTran' :: [Int] -> [BS.ByteString] -> Tran
parseTran' ks ws = Tran a k p
             where pre = (ks !! 0) - 1
                   rem = drop pre ws
                   a = BS.unwords $ take pre ws
                   k = BS.unwords $ take (length ks) rem
                   p = BS.unwords $ drop (length ks) rem

data Keys   = Keys [Int] | Error String
data Master = Master BS.ByteString BS.ByteString deriving Show -- keys and values
data Tran   = Tran BS.ByteString BS.ByteString BS.ByteString deriving Show -- values and keys and values
data OutTran = OkTran BS.ByteString | NgTran BS.ByteString deriving Show

keys = string "key=" >> (try(rangekey) <|> try(singlekey) )

singlekey = do n <- many1 digit
               return $ Keys [read n::Int]

rangekey = do n <- many1 digit
              char '/'
              m <- many1 digit
              return $ Keys [(read n::Int)..(read m::Int)]
